### PR TITLE
fix: split out sdk init

### LINKED ISSUE #6621: dagger functions/call/etc should fail if none present
@kpenfound noticed this fun one while migrating to new the `source` fields:

```
❯ dagger init --sdk=go --name=playground --source .
Initialized module playground in .
❯ ls
dagger.gen.go  dagger.json  go.mod  go.sum  main.go  querybuilder
❯ rm main.go
❯ dagger functions
Name             Description
container-echo   example usage: "dagger call container-echo --string-arg yo stdout"
grep-dir         example usage: "dagger call grep-dir --directory-arg . --pattern GrepDir"
```

Even though there's *no* code in `main.go`, we still get some default generated functions - I assume this is due to changes in the auto-generation code, to always apply.

### LINKED ISSUE #7834: Runtime modules need to know if a module is being created
The Python SDK tries to detect if a module is being created or not, in order to not create files from the template otherwise. Not doing this leads to:
 - https://github.com/dagger/dagger/issues/6621

We don’t have a reliable way to know if sources are being created for the first time. Here’s what Python does today:
- When there’s no `dagger.json`, it’s definitely a new module.
- If there’s a `dagger.json` but no `pyproject.toml` file (TS has `package.json`, but in certain situations Go may not have a `go.mod`), it’s a `dagger develop --sdk=` after an earlier `dagger init`.
- The SDK allows using a `pyproject.toml` for custom configuration, like choosing a different Python version for the module, even before `dagger init`. In this use case, there’s no way to tell if you do a `dagger develop --sdk=` after a `dagger init` because the CLI updates `dagger.json` with `"sdk"` before calling codegen.

We need to make sure that information (sources being created) is passed on to runtime modules reliably and update all the SDKs to only create new sources on initial creation. That solves https://github.com/dagger/dagger/issues/6621.

### PR DESCRIPTION
**Pull request by jedevc**
- Fixes #6621
- Fixes #7834
- Kind of generally part of #7707.

The idea is to more explicitly break down each SDKs initialization steps - we should only generate the base template in `Init`. This avoids the annoying implicit generation of `main.go` when no source files exist.

TODO:
- [ ] Implement for all the core SDKs
- [ ] Implement for the community SDKs and/or introduce a fallback if the SDK doesn't define `init`
- [ ] Cleanup

### CODE REVIEW

File: core/schema/sdk.go [hunk 1]
```diff
@@ -330,12 +334,38 @@ type goSDK struct {
 	dag  *dagql.Server
 }
 
+func (sdk *goSDK) Init(
```
[the remark this answers is not in the document]
Comment (jedevc) [reply core/schema/sdk.go:337]: The idea is that the codegen will call `init` + `codegen` for `dagger init`, but only `codegen` for `dagger develop`.

But we'll make sure to have compat fallbacks, so `init` won't be called if the module SDK doesn't implement it, so there's a nice upgrade path there.

[the remark this answers is not in the document]
Comment (helderco) [reply core/schema/sdk.go:337]: Should call init also on `dagger develop` if `--sdk` is used.

**Comment by helderco** [comment]
Maybe it should have a different name, so it's not implied that this function runs before all others. For example, you may need to generate some files depending on what files are there in the module. Perhaps the most flexible solution is for the engine to set an optional constructor argument for `init` and save it in the struct (like `sdkSourceDir`). 

This way you are free to use that value in any function. I would probably just do that if this was a function:
```go
func (m *MySdk) Initialize() *MySdk {
	m.init = true
	return m
}
```

**Comment by jedevc** [comment]
I think this should still be a separate function. We now have various `InitOpts` (e.g. this is how `embed` is propagated around). We should pass these through an `Initialize` function, instead of to codegen, which I think is more confusing.

SDKs that don't have anything special on `Initialize` can just do what you're suggesting.

> Maybe it should have a different name, so it's not implied that this function runs before all others.

Agreed, how about `Boilerplate`? Or `Template`? Or similar.

**Comment by helderco** [comment]
I've been hitching for this for a long time 🙂

> I think this should still be a separate function. […] SDKs that don't have anything special on `Initialize` can just do what you're suggesting.

I'm ok with that 👍

> Agreed, how about `Boilerplate`? Or `Template`? Or similar.

How about just `Init`, or `Bootstrap`?

`Initialize` is actually not a bad name considering that it means the target module is a new one, but the term is also overloaded since we use it in other places to mean "loading" the module (even an existing one), and SDK authors may think it also applies to existing modules, just something you’re supposed to do in the beginning of the loading process.

But `Init` matches more closely to `dagger init`, `git init`, `uv init`, etc.

> * [ ]  Implement for all the core SDKs

You can just create a function that does nothing, in Python and TypeScript. Me and Tom can either migrate in this PR or on a follow-up, but wouldn’t want you to get blocked on that.

> * [ ]  Implement for the community SDKs and/or introduce a fallback if the SDK doesn't define `init`

This would be ideal for any SDK actually, if the engine checks if a runtime module implements a given function. We’ll need this eventually in https://github.com/dagger/dagger/issues/7707. However, I’m ok with a quicker solution for just this case, until we work on that issue.

We also need to remember that there’s folks that have created a custom SDK extending the Python SDK ([see example](https://github.com/dagger/dagger/blob/ff2d7bf76a96a89226f0df28187ad823afc20bd8/core/integration/testdata/modules/python/git-dep/dagger.json#L3) in our test suite), in order to customize some aspect of it. I think primarily for corporate users needing private access to services. May be less necessary now for that use case though.

**Comment by jedevc** [comment]
Rebased, and chucked some huge simplifications to the go mod update and such in.

Not quite sure the final form of this, so keeping in draft, and discussing with @helderco out-of-band - this has a lot of connections to https://github.com/dagger/dagger/issues/7707.

**Pull request closed without merging**